### ISSUE TITLE
How can I set User Agent for Github login

### ISSUE THREAD

**Issue by dlednik** [SocialConnect/auth#108]
### Issue
I upgraded CakePHP to v4 and with that auth plugin gets upgraded too to version 3.1.

Is there a way to add User-Agent to the github login configuration?
After upgrade I get this error when trying to login:
```
2020-02-23 01:54:24 Error: [SocialConnect\Provider\Exception\InvalidResponse] API response with error code
Request URL: /social-auth/callback/github?code=c06d043cc8ea5fcd53b4&state=cb84af630eefaf5fe23e6e503df08efd
Referer URL: http://localhost:8765/
Provider Response: Request forbidden by administrative rules. Please make sure your request has a User-Agent header (http://developer.github.com/v3/#user-agent-required). Check https://developer.github.com for other possible causes.

Stack Trace:
#0 .../vendor/socialconnect/auth/src/Provider/AbstractBaseProvider.php(296): SocialConnect\Provider\AbstractBaseProvider->executeRequest(Object(Zend\Diactoros\Request))
#1 .../vendor/socialconnect/auth/src/OAuth2/Provider/GitHub.php(78): SocialConnect\Provider\AbstractBaseProvider->request('GET', 'user', Array, Object(SocialConnect\OAuth2\AccessToken))
#2 .../vendor/admad/cakephp-social-auth/src/Middleware/SocialAuthMiddleware.php(275): SocialConnect\OAuth2\Provider\GitHub->getIdentity(Object(SocialConnect\OAuth2\AccessToken))
#3 .../vendor/admad/cakephp-social-auth/src/Middleware/SocialAuthMiddleware.php(216): ADmad\SocialAuth\Middleware\SocialAuthMiddleware->_getProfile('github', Object(Cake\Http\ServerRequest))
#4 .../vendor/admad/cakephp-social-auth/src/Middleware/SocialAuthMiddleware.php(178): ADmad\SocialAuth\Middleware\SocialAuthMiddleware->_handleCallbackAction(Object(Cake\Http\ServerRequest))
#5 .../vendor/cakephp/cakephp/src/Http/Runner.php(73): ADmad\SocialAuth\Middleware\SocialAuthMiddleware->process(Object(Cake\Http\ServerRequest), Object(Cake\Http\Runner))
#6 .../vendor/cakephp/cakephp/src/Routing/Middleware/RoutingMiddleware.php(156): Cake\Http\Runner->handle(Object(Cake\Http\ServerRequest))
#7 .../vendor/cakephp/cakephp/src/Http/Runner.php(73): Cake\Routing\Middleware\RoutingMiddleware->process(Object(Cake\Http\ServerRequest), Object(Cake\Http\Runner))
#8 .../vendor/cakephp/cakephp/src/Routing/Middleware/AssetMiddleware.php(68): Cake\Http\Runner->handle(Object(Cake\Http\ServerRequest))
#9 .../vendor/cakephp/cakephp/src/Http/Runner.php(73): Cake\Routing\Middleware\AssetMiddleware->process(Object(Cake\Http\ServerRequest), Object(Cake\Http\Runner))
#10 .../vendor/cakephp/cakephp/src/Error/Middleware/ErrorHandlerMiddleware.php(118): Cake\Http\Runner->handle(Object(Cake\Http\ServerRequest))
#11 .../vendor/cakephp/cakephp/src/Http/Runner.php(73): Cake\Error\Middleware\ErrorHandlerMiddleware->process(Object(Cake\Http\ServerRequest), Object(Cake\Http\Runner))
#12 .../vendor/cakephp/debug_kit/src/Middleware/DebugKitMiddleware.php(60): Cake\Http\Runner->handle(Object(Cake\Http\ServerRequest))
#13 .../vendor/cakephp/cakephp/src/Http/Runner.php(73): DebugKit\Middleware\DebugKitMiddleware->process(Object(Cake\Http\ServerRequest), Object(Cake\Http\Runner))
#14 .../vendor/cakephp/cakephp/src/Http/Runner.php(58): Cake\Http\Runner->handle(Object(Cake\Http\ServerRequest))
#15 .../vendor/cakephp/cakephp/src/Http/Server.php(90): Cake\Http\Runner->run(Object(Cake\Http\MiddlewareQueue), Object(Cake\Http\ServerRequest), Object(App\Application))
#16 .../webroot/index.php(40): Cake\Http\Server->run()
#17 {main}
```
On old CakePHP version 3.8 auth plugin vas version 1.2 (if i remember correctly) and login was working.

### My env

PHP: 7.2.x

Library: 3.1

Provider: github

Provider options:

```php
[
    'options' => []
];
```
Will update some info later when I get to my machine.

**Comment by ovr** [comment]
Tested GitHub on demo site, https://sc.lowl.io/, it works (but it's working on own https://github.com/SocialConnect/http-client)

1. Related to `User-agent`, It's defined in `Http\Client` that you use

For cake plugin from @ADmad it's working on top of `Cacke\Client`

![image](https://user-images.githubusercontent.com/572096/75110161-b2243d00-563b-11ea-8000-a1286a1edfa8.png)

You should mutate the Client, but how I dont know

@ADmad can we add to configuration to pass another httpClient? (I saw this configuration but in `master` branch)

2. Now it's not possible to set another User-Agent in Provider's configuration, We needed to think, did we rly need it?

@ADmad @heiglandreas what do you think?

Thanks

**Comment by ADmad** [comment]
@ovr Your `\SocialConnect\HttpClient\Curl` http client class sets user agent through curl options hence you might have not encountered this error.

My plugin uses CakePHP's http client which doesn't set user agent by default which causes this issue.

IMO the request instance created by Github provider should have the `User-Agent` header to avoid this problem.

**Comment by ADmad** [comment]
@ovr I had typed previous response before I saw yours :slightly_smiling_face:

Setting the user agent in http client itself wouldn't be the right thing. What if someone else using `socialconnect/auth` uses another http client?  He wouldn't know that setting user agent is required for Github as I didn't.

Hence as I mentioned earlier it's the `Github` provider which should handle setting the `User-Agent` header. So I would go with option `2`. The provider can set can a default value if none is provided.

**Comment by ovr** [comment]
> The provider can set can a default value if none is provided.

1 if you are speaking about default user-agent in http-client

We dont know that: Did used HTTP client setup `User-Aget` or not 

For this case: there are only two variants for use: set or not to set.

if we will setup it in Provider, we should add an ability in configuration to change it, it's why it's more easy to do it on client side

2. If we will setup User\Agent, this will be not a real user agent

```
SocialConnect\Auth (https://github.com/SocialConnect/auth) v3
```

All HTTP\Clients will have same `User-Agent` for backend, backend will not able to diagnose errors if someone http client will have a bug

**Comment by ovr** [comment]
Sorry for late reply, workaround released in 3.1.3

@dlednik can you aprove that it's working now?

Thanks

**Comment by ovr** [comment]
I am going close this issue due unactivity and unapproved fix from library side.

Thanks!
